In [3]:
from SparseInteractions import SparseInteractions

ModuleNotFoundError: No module named 'SparseInteractions'

# From Raw Data to Predictions

This notebook is designed as a follow-up to the [Machine Learning with the Experts: School Budgets](https://www.datacamp.com/courses/machine-learning-with-the-experts-school-budgets) course on Datacamp. We won't explain all the tools and techniques we use here. If you're curious about any of the tools, code, or methods used here, make sure to check out the course!

In [1]:
%matplotlib inline
from __future__ import division
from __future__ import print_function

# ignore deprecation warnings in sklearn
import warnings
warnings.filterwarnings("ignore")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import os
import sys

# add the 'src' directory as one where we can import modules
src_dir = os.path.join(os.getcwd(), os.pardir, 'D:\DATA_SCIENCE\DrivenData\Box-Plots for Education\')
sys.path.append(src_dir)

sys.path.append('D:\GitHub\py.modules')       # Run if custom modules are not imported
                       
from data.multilabel import multilabel_sample_dataframe, multilabel_train_test_split
from SparseInteractions import SparseInteractions
from models.metrics import multi_multi_log_loss

# Load Data

First, we'll load the entire training data set available from DrivenData. In order to make this notebook run, you will need to: 
 - [Sign up for an account on DrivenData](http://www.drivendata.org)
 - [Join the Box-plots for education competition](https://www.drivendata.org/competitions/46/box-plots-for-education-reboot/)
 - Download the competition data to the `data` folder in this repository. Files should be named `TrainingSet.csv` and `TestSet.csv`.
 - Enjoy!

In [2]:
path_to_training_data = os.path.join(os.pardir,
                                     'data',
                                     'TrainingSet.csv')

df = pd.read_csv(path_to_training_data, index_col=0)
print(df.shape)

(400277, 25)


In [ ]:
print(df.describe())                                              # Print the summary statistics

plt.hist(df['FTE'].dropna())                                      # Create the histogram
plt.title('Distribution of %full-time \n employee works')
plt.xlabel('% of full-time')
plt.ylabel('num employees')
plt.show()

In [ ]:
df.dtypes.value_counts()

# Resample Data

400,277 rows is too many to work with locally while we develop our approach. We'll sample down to 40,000 rows so that it is easy and quick to run our analysis.

We'll also create dummy variables for our labels and split our sampled dataset into a training set and a test set.

In [3]:
LABELS = ['Function',
          'Use',
          'Sharing',
          'Reporting',
          'Student_Type',
          'Position_Type',
          'Object_Type', 
          'Pre_K',
          'Operating_Status']

NON_LABELS = [c for c in df.columns if c not in LABELS]

SAMPLE_SIZE = 40000

sampling = multilabel_sample_dataframe(df,
                                       pd.get_dummies(df[LABELS]),
                                       size=SAMPLE_SIZE,
                                       min_count=25,
                                       seed=43)

dummy_labels = pd.get_dummies(sampling[LABELS])

X_train, X_test, y_train, y_test = multilabel_train_test_split(sampling[NON_LABELS],
                                                               dummy_labels,
                                                               0.2,
                                                               min_count=3,
                                                               seed=42)

In [ ]:
""" Convert to Category """ 
# categorize_label = lambda x: x.astype('category')                 # Define the lambda function: categorize_label
# df[LABELS] = df[LABELS].apply(categorize_label, axis=0)           # Convert df[LABELS] to a categorical type
# print(df[LABELS].dtypes)                                          # Print the converted dtypes

In [ ]:
""" Calculate number of unique values for each label """
# num_unique_labels = df[LABELS].apply(pd.Series.nunique)

# num_unique_labels.plot(kind='bar')
# plt.xlabel('Labels')
# plt.ylabel('Number of unique values')
# plt.show()

In [ ]:
""" SPLITTING ALL DATA """
# from data.multilabel import multilabel_sample_dataframe, multilabel_train_test_split
# labels_to_use = pd.get_dummies(df[LABELS])
# X_train, X_test, y_train, y_test = multilabel_train_test_split(df[NON_LABELS], dummy_labels, 0.2, min_count=3, seed=42)

# Create preprocessing tools

We need tools to preprocess our text and numeric data. We'll create those tools here. The `combine_text_columns` function will take a DataFrame of text columns and return a single series where all of the text in the columns has been joined together.

We'll then create `FunctionTransformer` objects that select our text and numeric data from the dataframe.

Finally, we create a custom scoring method that uses the `multi_multi_log_loss` function that is the evaluation metric for the competition.

In [4]:
NUMERIC_COLUMNS = ['FTE', "Total"]

def combine_text_columns(data_frame, to_drop=NUMERIC_COLUMNS + LABELS):
    """ Takes the dataset as read in, drops the non-feature, non-text columns and then combines all of the text columns
        into a single vector that has all of the text for a row.
        
        :param data_frame: The data as read in with read_csv (no preprocessing necessary)
        :param to_drop (optional): Removes the numeric and label columns by default.
    """    
    to_drop = set(to_drop) & set(data_frame.columns.tolist())        # drop non-text columns that are in the df
    text_data = data_frame.drop(to_drop, axis=1)        
    text_data.fillna("", inplace=True)                               # replace nans with blanks
        
    return text_data.apply(lambda x: " ".join(x), axis=1)            # joins all text items in a row with a space between


In [5]:
from sklearn.preprocessing import FunctionTransformer

get_text_data = FunctionTransformer(combine_text_columns, validate=False)              # validate=False, don't check for NaNs
get_numeric_data = FunctionTransformer(lambda x: x[NUMERIC_COLUMNS], validate=False)

In [6]:
get_text_data.fit_transform(sampling.head(5))

38     OTHER PURCHASED SERVICES  SCHOOL-WIDE SCHOOL P...
70     Extra Duty Pay/Overtime For Support Personnel ...
198    Supplemental *  Operation and Maintenance of P...
209    REPAIR AND MAINTENANCE SERVICES  PUPIL TRANSPO...
614     GENERAL EDUCATION LOCAL EDUCATIONAL AIDE,70 H...
dtype: object

In [7]:
get_numeric_data.fit_transform(sampling.head(5))

FTE         Total
38    NaN    653.460000
70    NaN   2153.530000
198   NaN  -8291.860000
209   NaN    618.290000
614  0.71  21747.666875

In [8]:
from sklearn.metrics.scorer import make_scorer

log_loss_scorer = make_scorer(multi_multi_log_loss)

# Train model pipeline

Now we'll train the final pipeline from the course that takes text and numeric data, does the necessary preprocessing, and trains the classifier.

In [9]:
from sklearn.feature_selection import chi2, SelectKBest

from sklearn.pipeline import Pipeline, FeatureUnion

from sklearn.preprocessing import Imputer, MaxAbsScaler
from sklearn.feature_extraction.text import HashingVectorizer

from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.preprocessing import 

TOKENS_ALPHANUMERIC = '[A-Za-z0-9]+(?=\\s+)'                       # Create the token pattern

In [10]:
%%time

# set a reasonable number of features before adding interactions
chi_k = 300

# create the pipeline object
pl = Pipeline([
        ('union', FeatureUnion(
            transformer_list = [
                ('numeric_features', Pipeline([
                    ('selector', get_numeric_data),
                    ('imputer', Imputer())
                ])),
                ('text_features', Pipeline([
                    ('selector', get_text_data),
                    ('vectorizer', HashingVectorizer(token_pattern=TOKENS_ALPHANUMERIC,
                                                     non_negative=True, norm=None, binary=False,
                                                     ngram_range=(1, 2))),
                    ('dim_red', SelectKBest(chi2, chi_k))
                ]))
             ]
        )),
        ('int', SparseInteractions(degree=2)),
        ('scale', MaxAbsScaler()),
        ('clf', OneVsRestClassifier(LogisticRegression()))
    ])

# fit the pipeline to our training data
pl.fit(X_train, y_train.values)

# print the score of our trained pipeline on our test set
print("Logloss score of trained pipeline: ", log_loss_scorer(pl, X_test, y_test.values))
print('Accuracy score of trained pipeline: ', pl.score(X_test, y_test.values))

Logloss score of trained pipeline:  2.14831287426
CPU times: user 21min 28s, sys: 53.8 s, total: 22min 22s
Wall time: 6min 38s


# Predict holdout set and write submission

Finally, we want to use our trained pipeline to predict the holdout dataset. We will write our predictions to a file, `predictions.csv`, that we can submit on [DrivenData](http://www.drivendata.org)!

In [11]:
path_to_holdout_data = os.path.join(os.pardir,
                                    'data',
                                    'TestSet.csv')

# Load holdout data
holdout = pd.read_csv(path_to_holdout_data, index_col=0)

# Make predictions
predictions = pl.predict_proba(holdout)

# Format correctly in new DataFrame: prediction_df
prediction_df = pd.DataFrame(columns=pd.get_dummies(df[LABELS]).columns,
                             prefix_sep='__').columns, 
                             index=holdout.index,
                             data=predictions)


# Save prediction_df to csv called "predictions.csv"
prediction_df.to_csv("predictions.csv")

############ DRAFT ############

<b>Penalizing highly confident wrong answers</b><br>
<b>Log loss</b> provides a steep penalty for predictions that are both wrong and confident, i.e., a high probability is assigned to the incorrect class.<p>
Log loss penalizes highly confident wrong answers much more than any other type. This will be a good metric to use on your models.

In [ ]:
def compute_log_loss(predicted, actual, eps=1e-14):
    predicted = np.clip(predicted, eps, 1 - eps)
    loss = -1 * np.mean(actual * np.log(predicted)
                               + (1 - actual)
                               * np.log(1 - predicted))
    return loss

# compute_log_loss(predicted=1, actual=0.85)             #  
# compute_log_loss(predicted=0, actual=0.99)             # Wrong and Confidend

In [ ]:
### SPARSE INTERACTIONS
from itertools import combinations
import numpy as np
from scipy import sparse
from sklearn.base import BaseEstimator, TransformerMixin

class SparseInteractions(BaseEstimator, TransformerMixin):
    def __init__(self, degree=2, feature_name_separator="_"):
        self.degree = degree
        self.feature_name_separator = feature_name_separator

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        if not sparse.isspmatrix_csc(X):
            X = sparse.csc_matrix(X)

        if hasattr(X, "columns"):
            self.orig_col_names = X.columns
        else:
            self.orig_col_names = np.array([str(i) for i in range(X.shape[1])])

        spi = self._create_sparse_interactions(X)
        return spi

    def get_feature_names(self):
        return self.feature_names

    def _create_sparse_interactions(self, X):
        out_mat = []
        self.feature_names = self.orig_col_names.tolist()

        for sub_degree in range(2, self.degree + 1):
            for col_ixs in combinations(range(X.shape[1]), sub_degree):
                # add name for new column
                name = self.feature_name_separator.join(self.orig_col_names[list(col_ixs)])
                self.feature_names.append(name)

                # get column multiplications value
                out = X[:, col_ixs[0]]
                for j in col_ixs[1:]:
                    out = out.multiply(X[:, j])

                out_mat.append(out)

        return sparse.hstack([X] + out_mat)

In [ ]:
# NLP EXAMPLE
from sklearn.feature_extraction.text import CountVectorizer

TOKENS_BASIC = '\\S+(?=\\s+)'
TOKENS_ALPHANUMERIC = '[A-Za-z0-9]+(?=\\s+)'           # Create the token pattern: TOKENS_ALPHANUMERIC
# df.Position_Extra.fillna('', inplace=True)           # Fill missing values in df.Position_Extra

vec_basic = CountVectorizer(token_pattern=TOKENS_BASIC)                     # Instantiate basic CountVectorizer
vec_alphanumeric = CountVectorizer(token_pattern=TOKENS_ALPHANUMERIC)       # Instantiate alphanumeric CountVectorizer

text_vector = combine_text_columns(df)                                      # Create the text vector
vec_basic.fit_transform(text_vector)                                        # Fit and transform vec_basic
vec_alphanumeric.fit_transform(text_vector)                                 # Fit and transform vec_alphanumeric

print("There are {} tokens in the dataset".format(len(vec_basic.get_feature_names())))
print("There are {} alpha-numeric tokens in the dataset".format(len(vec_alphanumeric.get_feature_names())))